File: cortex/enterprise/promotion_models.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal

from pydantic import BaseModel, Field


PromotionStatus = Literal["draft", "candidate", "reviewed", "promoted", "rejected"]
PromotionDecisionType = Literal["approve", "reject"]


class PromotionIssue(BaseModel):
    file: str
    field: str
    message: str
    severity: Literal["error", "warning", "info"] = "warning"


class PromotionCandidate(BaseModel):
    origin_id: str
    doc_type: str
    local_rel_path: str
    local_abs_path: str
    dest_rel_path: str
    fingerprint: str
    status: PromotionStatus = "candidate"
    issues: list[PromotionIssue] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class PromotionDecision(BaseModel):
    decision: PromotionDecisionType
    actor: str
    decided_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat(timespec="seconds")
    )
    reason: str | None = None


class PromotionRecordEvent(BaseModel):
    event: Literal["candidate", "reviewed", "promoted", "rejected"]
    at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat(timespec="seconds")
    )
    actor: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)


class PromotionRecord(BaseModel):
    """
    Append-only record describing the lifecycle of one promotable document.

    `origin_id` is the stable idempotency key (project + local path).
    `fingerprint` is the content fingerprint of the source markdown (normalized).
    """

    origin_id: str
    local_rel_path: str
    doc_type: str
    dest_rel_path: str
    fingerprint: str
    status: PromotionStatus
    created_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat(timespec="seconds")
    )
    updated_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat(timespec="seconds")
    )
    decision: PromotionDecision | None = None
    events: list[PromotionRecordEvent] = Field(default_factory=list)

    def touch(self) -> None:
        self.updated_at = datetime.now(timezone.utc).isoformat(timespec="seconds")

File: rust/crates/cortex-enterprise/src/promotion_models.rs
//! Puerto de `cortex.enterprise.promotion_models`: records append-only del
//! ciclo de promoción. El orden de campos serde = orden de declaración
//! Pydantic ⇒ `model_dump_json` byte-parity (compacto, unicode crudo).

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::clock::{default_now_string, Clock};
use crate::error::EnterpriseError;

macro_rules! string_status {
    ($name:ident { $($variant:ident => $value:literal),+ $(,)? }) => {
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
        pub enum $name { $(#[serde(rename = $value)] $variant),+ }
        impl $name {
            pub fn as_str(self) -> &'static str { match self { $(Self::$variant => $value),+ } }
        }
        impl PartialEq<&str> for $name {
            fn eq(&self, other: &&str) -> bool { self.as_str() == *other }
        }
        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                f.write_str(self.as_str())
            }
        }
    };
}

string_status!(PromotionStatus {
    Draft => "draft",
    Candidate => "candidate",
    Reviewed => "reviewed",
    Promoted => "promoted",
    Rejected => "rejected"
});
string_status!(PromotionDecisionType {
    Approve => "approve",
    Reject => "reject"
});
string_status!(PromotionEventKind {
    Candidate => "candidate",
    ReviewedEvent => "reviewed",
    Promoted => "promoted",
    RejectedEvent => "rejected"
});

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionIssue {
    pub file: String,
    pub field: String,
    pub message: String,
    #[serde(default = "default_warning")]
    pub severity: String,
}
fn default_warning() -> String {
    "warning".to_string()
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionCandidate {
    pub origin_id: String,
    pub doc_type: String,
    pub local_rel_path: String,
    pub local_abs_path: String,
    pub dest_rel_path: String,
    pub fingerprint: String,
    #[serde(default = "default_candidate")]
    pub status: String,
    #[serde(default)]
    pub issues: Vec<PromotionIssue>,
    #[serde(default)]
    pub metadata: BTreeMap<String, serde_json::Value>,
}
fn default_candidate() -> String {
    "candidate".to_string()
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionDecision {
    pub decision: PromotionDecisionType,
    pub actor: String,
    #[serde(default = "default_now_string")]
    pub decided_at: String,
    #[serde(default)]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionRecordEvent {
    pub event: PromotionEventKind,
    #[serde(default = "default_now_string")]
    pub at: String,
    #[serde(default)]
    pub actor: Option<String>,
    #[serde(default)]
    pub payload: BTreeMap<String, serde_json::Value>,
}

/// Record append-only del ciclo de vida de un documento promotable.
/// `origin_id` es la clave de idempotencia (proyecto + ruta local);
/// `fingerprint` es el fingerprint normalizado del markdown fuente.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromotionRecord {
    pub origin_id: String,
    pub local_rel_path: String,
    pub doc_type: String,
    pub dest_rel_path: String,
    pub fingerprint: String,
    pub status: PromotionStatus,
    #[serde(default = "default_now_string")]
    pub created_at: String,
    #[serde(default = "default_now_string")]
    pub updated_at: String,
    #[serde(default)]
    pub decision: Option<PromotionDecision>,
    #[serde(default)]
    pub events: Vec<PromotionRecordEvent>,
}

impl PromotionRecord {
    /// `touch()`: actualiza `updated_at` al instante del reloj.
    pub fn touch(&mut self, clock: &dyn Clock) {
        self.updated_at = crate::clock::isoformat_seconds(clock.now());
    }

    /// Serialización `model_dump_json()` (compacta).
    pub fn to_json_line(&self) -> Result<String, EnterpriseError> {
        serde_json::to_string(self).map_err(|e| EnterpriseError::Validation(e.to_string()))
    }
}
